pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
+shines: for structures that go beyond simple class distinctions.
+
+Although pattern matching is a concept that has been known and used
+for decades, we propose a re-interpretation that centres around the
+principle of 'duck typing' and builds on existing features of the
+Python language such as iterable unpacking.  Patterns adopt the syntax
+of parameters as far as possible and, to a somewhat lesser degree,
+that of targets in iterable unpacking.  In contrast to iterable
+unpacking, pattern matching is a 'conditional' feature that has to
+avoid side-effects, i.e. extracting elements from an abstract iterable
+(thus working with actual sequences instead) or assigning to non-local
+targets such as object attributes or container elements.  Overall, we
+followed the guiding principle that patterns be static templates for
+the structure and type of objects, i.e. patterns should depend as
+little as possible on the surrounding context or current values of
+variables (other than the subject to be matched, that is).
```

Comment:
I'm not sure I care that much about constant-ness. Essentially a (constant) value pattern is always just going to do "load value, compare subject using `==`", and a compiler that optimizes that based on e.g. `final` annotations (PEP 591) can do the same optimization in other contexts as well. (In fact we have to be careful not to link these too closely to literal patterns, since we're not going to switch to `is` when the value is None/True/False.)

Reply:
I am not sure we are talking about the same thing here.  This passage is meant to explain why pattern matching is not binding attributes or subscripts, say.  It is basically a rephrasing of the notion of declarative style.

Declarative programming is built on the idea that you describe what you want and let the compiler figure out how to best deliver it.  Let us perhaps briefly look back at one of the very first examples for pattern matching ever:
```
fact 0 = 1
fact n = n * fact (n-1)
```
This was actually not meant as a recursive function the way we might think about it today.  The idea is much rather that this defines two equation and then have the compiler figure out how to determine what `fact 5`, say, should be.  The compiler might well choose to solve these equations using an ordinary iterative loop (nowadays known as tail-call optimisation).  Anyway, the point is that we do not care how the compiler translates our equations.

Of course, such an approach works _only if the declarative elements do not have side effects_.  As soon as you have any side effects, then you need to specify the exact order in which things are evaluated and how they can affect each other.  So, the guideline mentioned here is to avoid side effects as much as is possible, giving the compiler room to figure out the details.

Reply:
In the end I think this may go back to the same root cause as our disagreement about the indentation of `else` -- you see the declarative template, I see the generated code. I like the declarative approach, but I can never forget that underneath it all sits the Python interpreter.

Reply:
Up to a certain degree, yes.  However, there is yet another side to my view: I like to think along the lines of "_what semantics or guarantees do we need so that the compiler can generate good code automatically_" :-).